File: Pythoncode/93.复原IP地址.py
class Solution:
    def restoreIpAddresses(self, s: str):
        res = []
        self.helper("", 0, s, res)
        return res

    def helper(self, ip, counter, s, res):
        if counter > 4:
            return
        elif counter == 4:
            if len(s) == 0:
                res.append(ip)
        else:
            if len(s) > 3 * (4 - counter):
                return
            for i in range(min(3, len(s))):
                if int(s[:i+1]) <= 255:
                    if len(s[:i+1]) > 1 and s[:i+1][0] == "0": #不能以0开头
                        continue
                    newip = ip[:]
                    newip += s[:i+1]
                    newcounter = counter
                    newcounter += 1
                    if newcounter < 4:
                        newip += "."
                    self.helper(newip, newcounter, s[i+1:], res)

File: Cppcode/93.复原IP地址.cpp
#include <vector>
#include <string>


using namespace std;

class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> res;
        helper(s, 4, "", res);
        return res;
    }
    void helper(string s, int un_fill_space, string cur_ip, vector<string>& res)
    {
        if (un_fill_space == 0 && s.empty())
        {
            res.push_back(cur_ip.substr(1));
        }
        else if (s.empty() || un_fill_space == 0)
        {
            return;
        }
        else if (s[0] == '0')
        {
            helper(s.substr(1), un_fill_space-1, cur_ip+"."+"0", res);
        }
        else
        {
            for (size_t si = 1; si <= 3 && si <= s.size(); si++)
            {
                if (stoi(s.substr(0, si))>255) break;
                helper(s.substr(si), un_fill_space-1, cur_ip+"."+s.substr(0, si), res);
            }
        }
    }
};